src/PostgREST/DbStructure.hs

```diff
@@ -747,3 +752,19 @@ getPgVersion = H.query () $ H.statement sql HE.unit versionRow False
   where
     sql = "SELECT current_setting('server_version_num')::integer, current_setting('server_version')"
     versionRow = HD.singleRow $ PgVersion <$> HD.value HD.int4 <*> HD.value HD.text
+
+fillSessionWithSettings :: [(Text, Text)] -> H.Session ()
+fillSessionWithSettings settings =
+    -- combine all environment-setting SQL queries into one Hasql.Session
+    sequence_ $ List.map buildSetConfigQuery settings
```

Comment:
This sequence_ call will generate a roundtrip to the db for each var resulting in a slower "first request". This will be especially painful for small deployments where there is no big stream of requests and the connection is released after 1-2 seconds so basically every request will be a slow "first request".
All the above queries can be done in a single roundtrip like we do in middleware
`H.sql $ toS.mconcat $ setSchemaSql ++ setRoleSql ++ claimsSql ++ headersSql ++ cookiesSql`
so the resulting string will be `select set_config ...; select set_config...;` or even better it can be
`select set_config(), set_config()`.
Ideally you would do it with a prepared statement like `select set_config($1, $2), set_config($3, $4) ...` if you can get hasql to run a dynamically generated prepared statement, if not, then just `H.sql` with the vars in the string directly (piped through our formatting functions) will be ok.

(everything else looks great)

Reply:
Ok, I'll see what I can do :)